archethic-foundation/archethic-node
# Origin transaction should verify existing public key#583

# Description
 
`/api/origin_key` -> Origin endpoint is leveraged to add origin public key and its corresponding certificate, An origin transaction is created, on post request, with content as origin_public_key and certificate, But there are no checks for to ensure that this origin public key has already been included in existing transaction. A  check is introduced leveraging the OriginKey store to check if the origin Public key already exists before creating/accepting the transaction.

Checks are done at two different location/Stages:
 - In the `origin Api controller`, During post call
 - In `Pending_transaction_validation.ex` in do_accept_tx(:origin), validation of tx

New error logging for both locations api call and during validation is introduced. 

##Fixes 
- [ ] #583

## Type of change
- Bug fix 
   - Check for existing Origin Public key
- New feature
    -  Check for existing Origin Public key
- This change requires a documentation update
   - New error reporting if origin public key already exists

# How Has This Been Tested?
- [x] mix dev.lbench regression
- [x] Running on multiple nodes
- [x] unit test cases 

### How to test?
- `key certificate is ""` -Crypto.get_key_certificate(public_key)

-  Accepting a Origin tx with no existing origin public key in store
    - Generate pbkey, cert and post an origin api call with base16 encoded public key and certificate
      ``` 
     iex>  alias Archethic.Crypto
     iex>  alias Archethic.SharedSecrets.MemTable.OriginKeyLookup
     iex>  {public_key, _} = Crypto.derive_keypair("has_origin_public_key_Seed", 0)
     iex>  certificate = Crypto.get_key_certificate(public_key)
     iex> pb_key = public_key|>Base.encode16
     iex> cert = certificate|>Base.encode16
      ```
  - Check tx address in Origin Explorer
  
-  Rejecting a Origin tx with existing origin public key in store
   - Add the generated origin public key, via api, or method call
      ``` 
      iex>    OriginKeyLookup.add_public_key(:software, public_key)
      ```
   - Api response will be ```error - public key exists```
# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules
